### PR TITLE
Feature - Add logic to allow-on-list, detect, and block IPFS cIDs

### PR DESCRIPTION
**Pull request by 409H**
This allows us to push IPFS cIDs to the blocklist and have them blocked by the eth-phishing-detect service. We only need to note the CIDs and not block the gateways.

Related issue: https://github.com/MetaMask/eth-phishing-detect/issues/19532

**Note:** This requires a change to either `setupProviderConnection` for [`origin` value](https://github.com/MetaMask/metamask-extension/blob/b41022a97c2d1758d38e2fd4227531aa076dc9cb/app/scripts/metamask-controller.js#L4767C16-L4767C42) to include the URL paths **or** `isOnPhishingList()` to not use `origin` value and [instead pass the full URL into the check](https://github.com/MetaMask/metamask-extension/blob/b41022a97c2d1758d38e2fd4227531aa076dc9cb/app/scripts/metamask-controller.js#L2466) due to some IPFS gateways to have the CID in a subdomain and some gateways have the CID in the path.

### CODE REVIEW

**Comment by NikitaVr** [comment]
I tested this locally for a sanity check and everything is working well

- tried adding new blocklist items
- checked that only those CIDs get blocked
- checked that other domains and variations of paths are not impacted.

**Review by AlexHerman1 — DISMISSED** [review]
LGTM.

**Review by AlexHerman1 — DISMISSED** [review]
LGTM.

**Review by AlexHerman1 — APPROVED** [review]
LGTM. again!

**Review by AlexHerman1 — DISMISSED** [review]
lgtm?

File: src/detector.js [hunk 1]
```diff
@@ -49,11 +49,21 @@ class PhishingDetector {
     return result
   }
 
-  _check (domain) {
+  _check (url) {
+    let domain = url;
+
+    try {
+      domain = new URL(url).host;
```
Comment (FrederikBolding) [inline src/detector.js]: What is this meant to catch? Why would we suddenly have invalid URLs passed in? 🤔

File: src/config.json [hunk 2]
```diff
@@ -45,6 +45,7 @@
     "satoshilabs.design"
   ],
   "blacklist": [
+    "QmUDBVyGwqKdSayk7kDKUaj9J41Ft1DWizcKUx5UmgMgGy",
```
Comment (FrederikBolding) [inline src/config.json]: Does the live version of MetaMask correctly handle this? Or at least correctly ignore it?

File: src/detector.js (hunk 1 above)
Comment (409H) [reply src/detector.js]: Current usage of this passes in a hostname only which will fail and throw a TypeError, so we catch an exception and rebuild with a assumed https protocol

Comment (FrederikBolding) [reply src/detector.js]: But if we are gonna make changes upstream to pass in the full URL, will this still be needed?

File: src/config.json (hunk 2 above)
Comment (409H) [reply src/config.json]: From testing the `check()` method, it does ignore the entry -- I haven't tested the `MetaMask/core` usage but that uses `PhishingDetector.check` directly so it _should_ be fine

File: src/detector.js (hunk 1 above)
Comment (409H) [reply src/detector.js:56]: As per our call, I will rework this section of the code and ping you later today

Comment (409H) [reply src/detector.js]: https://github.com/MetaMask/eth-phishing-detect/pull/58432/commits/7e999b46bb0b811ca4d026b2acf359f0d3bc958d made changes as we discussed on the call

File: src/detector.js [hunk 3]
```diff
@@ -174,3 +189,17 @@ function matchPartsAgainstList(source, list) {
...
+
+function matchStringAgainstList(needle, list) {
+  const listMap = new Map(list);
+  return listMap.has(needle) ? needle : null;
+}
+
+function ipfsCidRegex(startEndMatch = true) {
+  // regex from https://stackoverflow.com/a/67176726
+  let reg = "Qm[1-9A-HJ-NP-Za-km-z]{44,}|b[A-Za-z2-7]{58,}|B[A-Z2-7]{58,}|z[1-9A-HJ-NP-Za-km-z]{48,}|F[0-9A-F]{50,}";
+  if(startEndMatch) {
```
Comment (FrederikBolding) [inline src/detector.js]: Since we always use `ipfsCidRegex` with `false` could we remove this extra complexity or does it have a use elsewhere?

**Review by FrederikBolding — DISMISSED** [review]

File: src/detector.js (hunk 3 above)
Comment (409H) [reply src/detector.js]: It's used with `true` on the tests https://github.com/MetaMask/eth-phishing-detect/pull/58432/files#diff-b7a738f35866f9728b7c1f8d2cb769aaaf35e55c270a530a28c3e0ec7227bd60R75

**Review by FrederikBolding — DISMISSED** [review]

**Comment by lyen1688** [comment]
https://zksync-airdrop-ashen.vercel.app/

Hello, this is a tool I made for batch query of airdrops. It has been widely welcomed and used. It does not require linking a wallet. Why is it blocked?

Original text here: https://x.com/1999_eth/status/1800439119056949715

**Comment by 409H** [comment]
Once merged, we need to merge https://github.com/MetaMask/eth-phishing-detect/pull/67637

**Comment by samczsun** [comment]
Closing this as it was merged in core

**Pull request closed without merging**